README
WHAT IS THIS?
-------------

This is release of Dominion (version 1.06), a world simulation and
conquest game.  Dominion runs on every version of UNIX on which we
have tested it.

The dominion source code is available by anonymous ftp on
max.physics.sunysb.edu [129.49.21.100], in the directory
pub/dominion-version.  For example, to get version 1.06 you would
type:

	ftp -i max.physics.sunysb.edu		(or ftp -i 129.49.21.100)
	<log in as "anonymous", with any password you want>
	cd pub/dominion-1.06
	ls
	binary
	mget DOM*

The files you will retrieve are "shell archives", so after
uncompressing, you un-bundle them with the Bourne shell.

Send bug reports to:
	rosalia@max.physics.sunysb.edu

Send flames to /dev/tty :-).

Don't forget that this is copylefted free software.  Read the file COPYING.


HOW DO YOU COMPILE DOMINION?
----------------------------

Once you have unpacked all the files, you might be lucky and just be
able to type "make all" and "make new-world".  Even better would be
"gmake" and "gmake new-world" if you have GNU Make.  If not:

You should browse the Makefile, and modify the top part of it.  There are
some customization options:

0.  if you don't have GNU make, remove the 2 lines that refer to RCS.
    otherwise, you can always type "make" with some target, such
    as "make all" or "make install" or "make new-world".

1.  if you have Berkeley UNIX, or sunOS, you should put -DBSD in the
    CFLAGS, rather than -DSYSV.  We have noticed that on mach (NeXT)
    and ULTRIX (VAX, DEC 5000) we need -DBSD, whereas AIX and System V
    will compile with -DSYSV.

2.  you should also set the LIBDIR (essential), THIS_GAME_LIBDIR
    (typically equal to LIBDIR), and BINDIR (not so essential).  You
    should also set the DEBUGFLAGS to "-O" if you want your compiler to
    optimize.  when dominion is shipped, the "-g" flag is specified.

3.  get the name of a basic mail program on your system, and set the
    Makefile variable DEFAULT_MAIL to be that.  This can be overruled
    with the DOMINION_MAIL environment variable, or in the dominion
    [O]ptions menu.

4.  do the same as step 3, with a basic visual editor:  the Makefile
    variable DEFAULT_EDITOR should be the editor most people
    use.  This can be overruled with the VISUAL environment variable,
    or in the dominion [O]ptions menu.

5.  There are some other Makefile variables that you might want to
    change, though they are not essential.  Some of the things we
    have in the Makefile are for software development.

    If you do not have LaTeXinfo, or you just want to use the manual
    as we shipt it, you should comment out the lines that generate
    dominion-man.info.  If you have LaTeXinfo, then you should set
    EMACS to be the name of GNU emacs on your system, and
    LTI_START to the name of the file with your LaTeXinfo lisp
    code.

    There is a newsgroup in which general things are posted.  This is
    called "News" by default, but it can be changed with the NEWS_GROUP
    variable in Makefile.

6.  Run "make all", and it will make 5 programs:
      "dom_make", "dom_add", "dominion", "dom_update" and "dom_print".

7.  Make sure that the file "dominion.info" is there.  We provide
    a copy of it with the distribution, in case you don't have the
    emacs lisp code needed to generate this file from "dominion.tex".
    This emacs lisp code comes with the LaTeXinfo package.

8.  Run "make new-world" to make a new world, then run "dom_add" to add
    a nation, then run "dominion" to play that nation, and
    "dom_update" to update the world after you run a turn etc...  It
    is convenient to run "make new-world" instead of manually typing
    "dom_make", since "make new-world" will copy all the relevant files
    to the LIBDIR.

9.  If you want to play with NPCs (nations run by the computer),
    type "dom_add -f npcs", where "npcs" is the name of the file
    with all information on these Non Player Countries.  From version
    1.05 on, the computer playes a pretty good game for the NPCs:
    they will definitely be a challenge!!  If you want to put in
    your own file for NPCs, just copy it to $LIBDIR/misc.

    The "npcs" file we distribute has 26 NPCs in it.  If you are
    running a small world, you might want to comment some of them
    out: do so by editing the file $LIBDIR/misc/npcs putting a '#'
    at the beginning of the line for each NPC you don't want.

10. If you don't like some of the hard-coded parameters, you can
    change them by editing the "dominion.h" and "costs.h" files.
    Other parameters are defined in "techno_levels", "army_types",
    "spirit_types", "npcs", and any file that starts with "mag_".
    If you change "dominion.h" or "costs.h", re-compile.  If you
    have modified the other files, just run "make lib-files".

11. To make a hardcopy of the manual dominion.tex, you must run LaTeX
    with the LaTeXinfo document style.  If you do not have this, you
    can get a copy of the two files "latexinfo.sty" and "tabular.sty"
    by anonymous ftp from
        max.physics.sunysb.edu [129.49.21.100]
    in the directory pub/dominion.  If you don't want to bother running
    LaTeX, you can find (in the same ftp directory) the files
    dominion.dvi.Z and dominion.PS.Z, and if you need it we can provide
    files ready for other laser printers.

12. We also supply a Game Master manual in the file "gm.tex".  This is
    formatted with ordinary LaTeX, and you can just use latex to format
    it and print it out.  This manual is just an initial sketch.

WHAT IS IT ALL ABOUT?
---------------------

This is a world simulation game developed by students at SUNY at Stony
Brook.  The intention is to write a game which offers the role playing
potential and complexity of relations found in conquer (written by Ed
Barlow, then picked up by Adam Bryant).  The plan is to extend the
game by offering general descriptions of many features (such as races,
magic spells, armies, spirits and technology powers).  For example,
the races of the world are not limited to 4 hard-coded races, but are
described in a file which can be modified at any point in the game by
the Game Master.  The same goes for army and spirit types.

Dominion is *not* another version of conquer; things work quite
differently, and we have *never* looked at the conquer source.  Still,
we acknowledge this most creative game which gave us our inspiration.
In September 1990, the author of conquer version 5 has heard of many
of the features of dominion, and has included them in conquer version
5.  He has acknowledged that these ideas come from Dominion.

The Game Master is given a nation, but this nation is sort of
"fragile", since it has no sectors, and such stuff.  We will make it
more solid, but meanwhile it is to be used mostly for its [E] command,
which allows the Game Master (who logs into the game as "Gamemaster")
to change the properties of a sector or of a nation.

Dominion development started in the spring of 1990, when we were
running a game of conquer at Stony Brook.  After the data file kept
crashing, and a whole lot of inconsistencies were discovered, and we
found that the source was too complex to be fixed, we decided to write
our own replacement.  Ed Barlow did a great job with conquer, and it
is a great game, but it was his first C program, and people who added
to it kept the endless case statements, and the special cases, and
worked on new features rather than re-writing.  The original name of
Dominion was Stony Brook World (sbw), but it has been named Dominion
since version 1.02.

One goal of dominion is that the ruler of a nation should always have
many choices available on how to invest resources, each one presenting
strong but distinct advantages.  Conquer contains both a good and a
bad example of this: you can invest metal in cities, ships or armies.
Either way you get advantages, and you have to choose.  On the other
hand, jewels are really not used for much else than getting magic
powers.  You can also use them to support monsters, but that is a
different order of magnitude, and few players get the monsters.  In
dominion, there should be several decisions you can make to invest all
your resources.

Kevin Hart has been working on the NPC code which allows the computer
to play various nations.  Starting with version 1.05, the
computer-played NPCs have begun posing a real challenge in the game.


ABOUT THE DEVELOPERS
--------------------

People who have written code or documentation so far are:

Mark Galassi (rosalia@dirac.physics.sunysb.edu)
Mike Fischer (greendog@max.physics.sunysb.edu)
Doug Novellano (doug@max.physics.sunysb.edu)
Keith Messing (keith@max.physics.sunysb.edu)
Alan Saporta (gandalf@max.physics.sunysb.edu)
Joanne Rosenshein (raven@max.physics.sunysb.edu)
Stephen Bae (sbae@max.physics.sunysb.edu)
Chris Coligado (noel@max.physics.sunysb.edu)
Stephen Underwood (su11+@andrew.cmu.edu)
Kevin Hart (hart@cs.andrews.edu)
C. Titus Brown (brown@dirac.physics.sunysb.edu)
Charles Ofria (charles@max.physics.sunysb.edu)

We keep in touch electronically and meeting in the Stony Brook
Institute for Theoretical Physics.  We also meet once a week for 1
hour, and some times have parties on Saturdays.  The tuesday meetings
are strictly for discussing code, the saturday parties are for
fantasizing about what we would like to do.

There is also a creative team of D&D and war-game players who are
giving suggestions on how magic, technology and battle should work.
We also hope to use some simple econometric models to calculate
parameters of the economy.


ABOUT THE CODE
--------------

There are 5 programs: dom_make, dom_add, dominion dom_update and
dom_print.  These all share some files, such as misc.c, nation.c and
file.c.

It would be nice to get, some day, some real graphical interfaces for
widely available personal computers (Amiga, Atari ST, Macintosh, IBM
PC) and windowing systems such as X and NeWS.  But for now we only
have the curses interface, and this is supported as the common
denominator.  There exist, of course, curses implementations for all
sorts of personal computers.


ABOUT SUPPORT AND MAINTAINANCE
------------------------------

We will definitely try to fix any bugs that are reported.  When
reporting a bug you should mention on which version of UNIX you found
it, and be precise.  A fix would be nice.

As for new ideas, we intend to move slowly in that area because
accepting innovations indiscriminately can be harmful in the long run.
Any idea which does not make the game more complicated is welcome, and
will be considered.  If you make your own modifications to the game,
remember to do so in accordance with the GNU General Public Licence
(see the file COPYING).


WHAT HAS BEEN DONE SO FAR?
--------------------------

We feel that the game is complete as it is, though lots of things
could be improved.  There is a file called TASKS with a list of
jobs for the enterprising hacker who wants to help with Dominion.


ChangeLog
Thu Jan 30 16:00:06 1992  Mark Galassi  (rosalia at max)

	* root.c: now root windows are cleared when created on BSD
	systems.

	* cur_stuff.c: fixed problem with return value of wget_number()
	function and friends.

	* spells.c: fixed a bug with the merge spell which allowed you to
	steal other nations' population by merging on hostile sectors.

	* mail.c and commands.c: fixed problem with permissions/uids
	etc...  now sending mail works both when forwarding and when
	sending locally.

Tue Jan 28 22:58:44 1992  Douglas J. Novellano  (doug at max)

	* c_news.c: write_newsrc(): Fixed (?) the .tmprc.n error bug.
	               Changed mv system call to cp and rm.

Mon Jan 27 04:30:09 1992  C. Titus Brown  (brown at dirac)

	* commands.c: Fixed aforementioned bug

	* c_news.c: Fixed the previously mentioned major bug

	* mail.c: Fixed the major bug with the init_screen and edit()

	* int_mail.c: Now, when the mail reader deletes all the messages
		      in the mailbox, the change is kept
		      (formerly, it tried to copy the deleted file)

	* int_mail.c: Fixed the major bug with the init_screen and
		      edit().

Sat Jan 25 18:59:55 1992  C. Titus Brown  (brown at dirac)

	* army.c: added routine "statline2_prompt" to all the relevant places

	* army.c: fixed problem with army status' running over into bonus'
		  in zoom_army.

	* misc.c: added routine "statline2_prompt" 

Fri Jan 24 02:30:43 1992  C. Titus Brown  (brown at dirac)

	* mktables.c: checked in (not in makefile)

	* dominion.tex: added in new army_types table (see mktable)

	* ext.c: changed description for army_examine, spelling, etc.

Thu Jan 23 23:36:33 1992  C. Titus Brown  (brown at dirac)

	* nation.c: new_army_type now checks to see if the army_type
		    being added already is there.

        * armylib.c: add_army_type now checks to see if the army_type
		     being added is already there.

Thu Jan 23 18:03:05 1992  Michael D Fischer  (greendog at max)

	* ext.c: changed some <a>rmy <e>xamine descriptions

Thu Jan 23 15:46:41 1992  Stephen H. Underwood  (heechee at max)

	* misc.h: removed extern declarations of index/strchr

	* army.c: added typecast to remove warning.

Fri Jan 17 13:50:42 1992  Stephen H. Underwood  (heechee at max)

	* army.h, spelllib.c: Fixed the problem with spells & merge
	so it is not possible to split or merge armies or spritis
	which are currently being affected by a spell.

Tue Jan 14 16:40:01 1992  Stephen H. Underwood  (heechee at max)

	* dominion.h, cur_stuff.c, Makefile: Added supporrt for PMAX, 
	fixing bug in the curses desplay for ultrix

	* dominion.c, printmap.c: Added support for ANDREW style
	authenticaion (since setuid/setgid doesn't function on AFS)

	* army.c, ext.c, menus.c, nation.c, addnation.c, file.c,
	amry.h, update.c, mail.c, user.c, spelllib.c
	Remove the slew of meaningless comments I left lying around.

Mon Jan 13 23:04:10 EST 1992 Stephen Underwood (heechee at max)

	* army.c: removed direct reference to strchr that would not
	compile under BSD.  

	* misc.h: Made sure that strchr or index were defined correctly
	as a proper fix to the above problems.

Fri Jan 10 15:58:17 1992  Michael D Fischer  (greendog at max)

	* printmap.c: Changed scaling from c-code to ps-code, and
	added axes to borders of map.

Tue Jan  7 22:36:35 1992  Michael D Fischer  (greendog at max)

	* printmap.c: Changes back to real userid to be able to write
	to user's directory.  Also a few minor changes.

Tue Jan  7 14:09:51 1992  Mark Galassi  (rosalia at max)

	* Addnation: now doesn't infinite-loop if there is no
	space of a new nation.  Behaves well on this count.

Fri Dec 13 08:11:54 1991  Mark Galassi  (rosalia at max)

	* Makefile: added the ChangeLog in the distribution

	* Makefile: now compiling as version 1.06beta

	* cinfo.c: fixed bug that find_info_file() was not returning
	the file pointer as it should have been. this caused core
	dumps on suns, and worked by miracle on MIPS/SGI.

	* armylib.c: fixed (finally) bug in army_is_in_sector() which
	caused all sorts of trouble with armies not being properly
	inserted into sectors.


spy.c
  /* spy.c -- things having to do with "other" nations and espionage */

/*
 * Copyright (C) 1990 Free Software Foundation, Inc.
 * Written by the dominion project.
 *
 * This file is part of dominion.
 *
 * dominion is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License as published
 * by the Free Software Foundation; either version 1, or (at your option)
 * any later version.
 *
 * This software is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this software; see the file COPYING.  If not, write to
 * the Free Software Foundation, 675 Mass Ave, Cambridge, MA 02139, USA.
 */

#include "dominion.h"
#include "misc.h"
#include <stdio.h>
#include <ctype.h>
#include <math.h>

  /* these macros represent the difficulty in gathering
     inteligence in various areas.
   */
#define SPY_POP 1
#define SPY_ECO 3
#define SPY_MAG 4
#define SPY_MIL 5
#define SPY_TECHNO 5
#define SPY_CAP 6

extern Sworld world;
extern Suser user;
extern char help_tag[];

  /* info on all nations in the world */
char nations_report()
{
  WINDOW *w;
  FILE *diplock;
  char s[200];
  Snation *np;
  char c;
  int i, done = 0, first_shown, n_shown, id;

  w = newwin(LINES-2, COLS, 0, 0); /* full screen */
  werase(w);
  touchwin(w);
  sprintf(s, "report on all nations");
  wmove(w, 0, (COLS-strlen(s))/2); /* make the string centered */
  wstandout(w);
  waddstr(w, s);
  wclrtoeol(w);
  wstandend(w);

    /* here go the guts */
  sprintf(s,"World size is: %dx%d, there are %d nations",
	    world.xmax, world.ymax, world.n_nations);
  wmove(w, 1, (COLS-strlen(s))/2); /* make the string centered */
  waddstr(w, s);
  mvwaddstr(w, 3, 0, " id");
  mvwaddstr(w, 4, 0, "---");
  mvwaddstr(w, 3, 5, "nation");
  mvwaddstr(w, 4, 5, "------");
  mvwaddstr(w, 3, 25, "mark");
  mvwaddstr(w, 4, 25, "----");
  mvwaddstr(w, 3, 31, "leader");
  mvwaddstr(w, 4, 31, "------");
  mvwaddstr(w, 3, 47, "race");
  mvwaddstr(w, 4, 47, "----");
  if (user.id == 0) {
    mvwaddstr(w, 3, 55, "money");
    mvwaddstr(w, 4, 55, "-----");
    mvwaddstr(w, 3, 65, "civil");
    mvwaddstr(w, 4, 65, "-----");
  }
    /* figure out which nations to show */
  first_shown = 1;
  n_shown = min(world.n_nations, LINES-10);
  while (!done) {
    strcpy(help_tag, "Nations Report");
    for (i = 0; i < n_shown && i+first_shown < world.n_nations; ++i) {
      np = &(world.nations[i+first_shown]);
      mvwprintw(w, i+5, 0, "%3d", np->id);
      wclrtoeol(w);
      mvwaddstr(w, i+5, 5, np->name);
      mvwaddch(w, i+5, 26, np->mark);
      sprintf(s,"%-12.12s",np->leader);
      mvwaddstr(w, i+5, 31, s);

      mvwaddstr(w, i+5, 47, np->race.name);

      if (!is_active_ntn(np)) {
	mvwaddstr(w, i+5, 60, "DESTROYED");
      } else if (user.id == 0) {
	mvwprintw(w, i+5, 55, "%d", np->money);
	mvwprintw(w, i+5, 65, "%d", get_n_civil(np));
      }
      if (np->npc_flag) {
	mvwaddstr(w, i+5, 75, "npc");
      }
    }
    wclrtobot(w);
    mvwaddstr(w, LINES-4, 4,
   "Options: [s]py on a nation, [<]/[,] previous screen, [>]/[.] next screen");
    mvwaddstr(w, LINES-3, 4,
	      "Reports: [b]udget, [p]roduction, [i]nfo, [d]iplomacy");
    wrefresh(w);
    statline("type space when done, or F to dump to a file", "nations_report");

    switch (c = getch()) {
    case '>':
    case '.':
      if (first_shown+n_shown < world.n_nations) {
	first_shown += n_shown;
      }
      break;
    case '<':
    case ',':
      if (first_shown > 1) {
	first_shown -= n_shown;
      }
      break;
    case 's':
      mvwaddstr(w, LINES-3, 4, "  Number of nation to spy on? ");
      wclrtoeol(w);
      if (wget_number(w, &id) > 0 && is_active_ntn(&world.nations[id])
	  && id != user.np->id) {
	if (user.id == 0) {	/* for the game master, give the total info */
	  (void) info_report(&world.nations[id]);
	} else {
	  spy_report(id);
	}
	touchwin(w);
      }
      break;
    case 'F':
      dump_current_screen(w, "nations_report");
      break;
    case ' ':
    case 'b':
    case 'p':
    case 'i':
    case 'd':
      done = 1;
      break;
    case '?':
      online_info();
      break;
    default:
      break;
    }
  }
  delwin(w);
  return c;
}

  /* allow a nation to spy on anther */
spy_report(id)
     int id;
{
  WINDOW *spyw;
  Snation *spied_np = &world.nations[id], *spying_np = user.np;
  char s[EXECLEN];
  char c;
  int done = 0, bribe;		/* amount of jewels to get info */
  int x, y;			/* for capital locations */

  strcpy(help_tag, "Nations Report");
  spyw = newwin(18, 60, LINES-22, (COLS-60)/2);
  werase(spyw);
  touchwin(spyw);
  while (!done) {
    sprintf(s, "Espionage Report on nation %s", spied_np->name);
    wmove(spyw, 1, (60-strlen(s))/2);
    wstandout(spyw);
    waddstr(spyw, s);
    wclrtoeol(spyw);
    wstandend(spyw);
    /* put the guts between here and the wrefresh() */
    mvwaddstr(spyw, 3, 5, "Spy on: [p]opulation, [e]conomy");
    mvwaddstr(spyw, 4, 5, "        [m]ilitary, ma[g]ic, [C]apital location");
    mvwaddstr(spyw, 5, 5, "        [t]echology, [T]echnology theft");
    wclrtobot(spyw);
    box(spyw, '|', '-');
    wrefresh(spyw);
    statline("type space when done, or F to dump to a file", "spy_report");
    c = getch();
    if (strchr("pemgCt", c) != NULL) {
      mvwaddstr(spyw, 10, 1,
		"How many jewels do you want to pay in bribes? ");
      if (wget_number(spyw, &bribe) < 1 || bribe <= 0) {
	continue;
      }
      if (bribe > spying_np->jewels) {
	statline2_err("Hit space", "You don't have enough jewels");
	continue;
      }
    }
    switch (c) {
    case '?':
      online_info();
      break;
    case ' ':
      done = 1;
      break;
    case 'p':			/* info on their population */
      mvwprintw(spyw, 12, 6, "Population is %d",
		spy_figure(get_n_civil(spied_np), bribe,
		spying_np, spied_np, SPY_POP) );
      break;
    case 'e':			/* info on their military */
      mvwprintw(spyw, 12, 4, "Money: %d",
		spy_figure(spied_np->money, bribe,
		spying_np, spied_np, SPY_ECO) );
      mvwprintw(spyw, 12, 20, "Jewels: %d",
		spy_figure(spied_np->jewels, bribe,
		spying_np, spied_np, SPY_ECO) );
      mvwprintw(spyw, 12, 36, "Metal: %d",
		spy_figure(spied_np->metal, bribe,
		spying_np, spied_np, SPY_ECO) );
      mvwprintw(spyw, 13, 4, "Food: %d",
		spy_figure(spied_np->food, bribe,
		spying_np, spied_np, SPY_ECO) );
      mvwprintw(spyw, 13, 20, "Tax: %d",
		spy_figure(spied_np->taxes, bribe,
		spying_np, spied_np, SPY_ECO) );
      break;
    case 'C':
      x = spy_figure(spied_np->capital.x, bribe, spying_np, spied_np, SPY_CAP);
      y = spy_figure(spied_np->capital.y, bribe, spying_np, spied_np, SPY_CAP);
      x = xrel(x, y, spying_np->capital);
      y = yrel(x, y, spying_np->capital);
      mvwprintw(spyw, 13, 4, "Capital is at (%d, %d) ", x, y);
      break;
    case 't':			/* info on their technology */
      mvwprintw(spyw, 12, 4, "Techno skill: %d",
		spy_figure(spied_np->tech_skill, bribe,
		spying_np, spied_np, SPY_TECHNO) );
      break;
    case 'm':			/* info on their military */
      mvwprintw(spyw, 12, 4, "Soldiers: %d",
		spy_figure(get_n_soldiers(spied_np), bribe,
		spying_np, spied_np, SPY_MIL) );
      break;
    case 'g':			/* info on their magic */
      mvwprintw(spyw, 12, 4, "Magic skill: %d",
		spy_figure(spied_np->mag_skill, bribe,
		spying_np, spied_np, SPY_MAG) );
      mvwprintw(spyw, 13, 4, "Spell pts.: %d",
		spy_figure(spied_np->spell_pts, bribe,
		spying_np, spied_np, SPY_MAG) );
      break;
    }
        /* this section is common to all bribes */
    if (strchr("pemgCt", c) != NULL) {
      mvwaddstr(spyw, 14, 10, "Hit space");
      wrefresh(spyw);
      get_space();
      spying_np->jewels -= bribe;
      cjewels(spying_np, -bribe);
    }
  }
  delwin(spyw);
}

spy_figure(n, expend, spying_np, spied_np, cost_fact)
     int n,			/* number we modify here */
       expend;			/* jewels spent on bribes */
     Snation *spying_np, *spied_np; /* nations involved */
     int cost_fact;		/* cost factor for type of info */
{
  double accuracy, error;
  int figure;			/* the figure we actually return */
  char s[400];

  if (spied_np->secrecy == 0) {
    spied_np->secrecy = 1;	/* avoid divide-by-zero */
  }
  accuracy =
    ((double) expend*spying_np->spy)/(cost_fact*spied_np->secrecy*100);
  
    /* now get the percent error */
  error = 100*( exp(-sqrt(accuracy)) + 1.0/(5.0+accuracy) );
  if (error > 100) {
    error = 100;
  }
/*  sprintf(s, "acc=%f,%%err=%f", accuracy, error); */
    /* now get the absolute error */
  error = 1 + n*error/100.0;
  figure = n + (RND() % ((int) error)) - (int) (error/2);

/*  statline2(s, ""); */

  return figure;
}
